internal/logger/telemetry_store_watermill.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logger
+
+import (
+	"fmt"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/rs/zerolog"
+
+	"github.com/stacklok/minder/internal/engine"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// TelemetryStoreWMMiddleware is a Watermill middleware that
+// logs the relevant telemetry data.
+type TelemetryStoreWMMiddleware struct {
+	l *zerolog.Logger
+}
+
+// NewTelemetryStoreWMMiddleware returns a new TelemetryStoreWMMiddleware.
+func NewTelemetryStoreWMMiddleware(l *zerolog.Logger) *TelemetryStoreWMMiddleware {
+	return &TelemetryStoreWMMiddleware{l: l}
+}
+
+// TelemetryStoreMiddleware is a Watermill middleware that
+// logs the relevant telemetry data.
+func (m *TelemetryStoreWMMiddleware) TelemetryStoreMiddleware(h message.HandlerFunc) message.HandlerFunc {
+	return func(msg *message.Message) ([]*message.Message, error) {
+		if err := m.Log(msg); err != nil {
+			return nil, fmt.Errorf("error logging telemetry: %w", err)
+		}
+
+		return h(msg)
+	}
```

Comment:
I think you want to pass the TelemetryStore along to be available for the rules engine, i.e.

```suggestion
	return func(msg *message.Message) ([]*message.Message, error) {
		ts := TelemetryStore{}
		logger := zerolog.Ctx(msg.Context())
		ctx := ts.WithTelemetry(logger.WithContext(msg.Context()))
		msg.SetContext(ctx)

		msgs, err := h(msg)
		
		// Record telemetry
		logMsg := logger.Info()
		if err != nil {
			logMsg = logger.Error()
		}
		ts.Record(logMsg)
		logMsg.Send()

		return msgs, err
	}
```

The "Record telemetry" part could partially go before the call to `h`, but I think we want to call `ts.Record` _after_ `h`, because the handler might want to store more telemetry that's part of the same record.